timvt/endpoints/factory.py

```diff
@@ -0,0 +1,169 @@
+
+    # FastAPI router
+    router: APIRouter = field(default_factory=APIRouter)
+
+    # TileMatrixSet dependency
+    tms_dependency: Callable[..., TileMatrixSet] = TileMatrixSetParams
+
+    # Table dependency
+    table_dependency: Callable[..., TableMetadata] = TableParams
+
+    # Database pool dependency
+    db_pool_dependency: Callable[..., Pool] = _get_db_pool
+
+    # Router Prefix is needed to find the path for /tile if the TilerFactory.router is mounted
+    # with other router (multiple `.../tile` routes).
+    # e.g if you mount the route with `/cog` prefix, set router_prefix to cog and
+    router_prefix: str = ""
+
+    def __post_init__(self):
+        """Post Init: register route and configure specific options."""
+        self.register_routes()
+
+    def register_routes(self):
+        """Register Tiler Routes."""
+        self.tile()
+        self.tilejson()
+        self.index()
+
+    def url_for(self, request: Request, name: str, **path_params: Any) -> str:
+        """Return full url (with prefix) for a specific endpoint."""
+        url_path = self.router.url_path_for(name, **path_params)
+        base_url = str(request.base_url)
+        if self.router_prefix:
+            base_url += self.router_prefix.lstrip("/")
+        return url_path.make_absolute_url(base_url=base_url)
+
+    ############################################################################
+    # /index
+    ############################################################################
+    def index(self):
+        """Register /index endpoint"""
+
+        @self.router.get(
+            "/index",
+            response_model=List[TableMetadata],
+            description="Return available tables.",
+            deprecated=True,
+        )
+        async def display_index(request: Request):
+            """Return JSON with available table metadata. """
+            return [
+                TableMetadata(
+                    **table, link=self.url_for(request, "tilejson", table=table["id"])
+                )
+                for table in request.app.state.Catalog
+            ]
+
+    ############################################################################
+    # /tiles
+    ############################################################################
+    def tile(self):
+        """Register /tiles endpoints."""
+
+        @self.router.get("/tiles/{table}/{z}/{x}/{y}.pbf", **TILE_RESPONSE_PARAMS)
+        @self.router.get(
+            "/tiles/{TileMatrixSetId}/{table}/{z}/{x}/{y}.pbf", **TILE_RESPONSE_PARAMS
+        )
+        async def tile(
+            z: int = Path(..., ge=0, le=30, description="Mercator tiles's zoom level"),
+            x: int = Path(..., description="Mercator tiles's column"),
+            y: int = Path(..., description="Mercator tiles's row"),
+            tms: TileMatrixSet = Depends(self.tms_dependency),
+            table: TableParams = Depends(self.table_dependency),
```

Comment:
leaving this here for now, but I'd argue that any sort of table or layer configuration should be pushed lower into the `VectorTileReader`.  This keeps layers decoupled from the API layer.